processor/detector.go

```diff
@@ -0,0 +1,26 @@
+package processor
+
+import (
+	"errors"
+	"strings"
+)
+
+// Given some content attempt to determine if it has a #! that maps to a known language and return the language
+func DetectSheBang(content string) (string, error) {
+	ProcessConstants()
+
+	if !strings.HasPrefix(content, "#!") {
+		return "", errors.New("Missing #!")
+	}
+
+	for k, v := range ShebangLookup {
+		for _, x := range v {
+			// detects both full path and env usage
+			if strings.Contains(content, "/"+x) || strings.Contains(content, " "+x) {
```

Comment:
This looks like it has the potential for false positives. The first case should be checking for the basename specifically, but this could match on directory names (e.g. `/python/perl/ruby` would match any of `python`, `perl` or `ruby` when only `ruby` should be matched.

The second case is only really valid if the first part is the path to an `env` program.

Reply:
That's a good point. Although... there is a special place in hell for those who organise their paths like that.

The second part is indeed for `env` detection.

Ill have a ponder on the first. I think the check might need to look for the first newline, then check if it has a postfix that matches. The `env` check would remain the same but within those same bounds.

Reply:
Although the problem is the fuzzyness of how people can use #!

Consider the following cases, which were all covered using the technique I did initially,

```
"#!/usr/bin/perl"
"#!  /usr/bin/perl"
"#!/usr/bin/perl -w"
"#!/usr/bin/env perl"
"#!  /usr/bin/env   perl"
"#!/usr/bin/env perl -w"
"#!  /usr/bin/env   perl   -w"
"#!/opt/local/bin/perl"
"#!/usr/bin/perl5"
```

I don't even know if the 3rd example `"#!/usr/bin/perl -w"` is legitimate actually. Anyway it did work with all of the above.

Perhaps for the first check removing all white-space and `-w` might be enough to resolve this.

Reply:
Updated with slightly modified code that meets all requirements and also deals with newlines. New test cases added.